### PR TITLE
Move 2pi scaling of zpk filters for better encapsulation and more explicit use

### PR DESCRIPTION
**Pull request by asouthgate**
2pi scaling happens by default inside bilinear_zpk for Hz -> rad/s. Extract that functionality so that i) bilinear transform is not also responsible for scaling (single responsibility principle) ii) it is used more explicitly where it is needed iii) the functions are more easily tested.

This MR will be followed by another MR that includes a comprehensive set of tests for filter design functionality, including zpk_transform, as well as bugfixes.

### CODE REVIEW

File: gwpy/signal/filter_design.py [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
 from scipy import signal
 
-from astropy.units import (Unit, Quantity)
+from astropy.units import (Quantity)
```
Comment (duncanmmacleod) [inline gwpy/signal/filter_design.py:31]:
```suggestion
from astropy.units import Quantity
```

File: gwpy/signal/filter_design.py [hunk 2]
```diff
@@ -256,7 +256,38 @@ def fir_from_transfer(transfer, ntaps, window='hann', ncorner=None):
...
+def convert_zpk_units(filt, unit):
+    """Convert zeros and poles created for a freq response in Hz to rad/s.
+
+    Parameters
+    ----------
+    filt : `tuple`
+        zeros, poles, gain
+
+    unit : `str`
+        `'Hz'` or `'rad/s'`
```
Comment (duncanmmacleod) [inline gwpy/signal/filter_design.py:268]:
```suggestion
        ``'Hz'`` or ``'rad/s'``
```

File: gwpy/timeseries/timeseries.py [hunk 3]
```diff
@@ -969,6 +969,10 @@ def zpk(self, zeros, poles, gain, analog=True, **kwargs):
             type of ZPK being applied, if `analog=True` all parameters
             will be converted in the Z-domain for digital filtering
 
+        unit: `str`
+            The frequency response units this filter was designed for
+             either Hz or rad/s. Default: 'Hz'.
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py:974]:
```suggestion
            The frequency response units this filter was designed for
            either Hz or rad/s. Default: 'Hz'.
```

File: gwpy/timeseries/timeseries.py [hunk 4]
```diff
@@ -986,7 +990,9 @@ def zpk(self, zeros, poles, gain, analog=True, **kwargs):
 
         >>> data2 = data.zpk([100]*5, [1]*5, 1e-10)
         """
-        return self.filter(zeros, poles, gain, analog=analog, **kwargs)
+
+        f = self.filter(zeros, poles, gain, analog=analog, unit=unit, **kwargs)
+        return f
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py:995]: Is this change necessary?

File: gwpy/timeseries/timeseries.py [hunk 5]
```diff
@@ -1080,11 +1090,21 @@ def filter(self, *filt, **kwargs):
         # parse filter
         form, filt = filter_design.parse_filter(filt)
 
+        unit = kwargs.pop('unit', False)
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py:1093]: `None` might be a more readable default for this case.

```suggestion
        unit = kwargs.pop('unit', None)
```

**Review by duncanmmacleod — CHANGES_REQUESTED** [review]
Some minor changes needed, but otherwise a nice change that I'll be happy to merge shortly. Thanks @asouthgate.

File: gwpy/timeseries/timeseries.py (hunk 4 above)
Comment (asouthgate) [reply gwpy/timeseries/timeseries.py:995]: I think it was a line length workaround.

Comment (duncanmmacleod) [reply gwpy/timeseries/timeseries.py:995]: Can this be reverted? Assigning to a variable is unnecessary here. If the line is too long, just wrap:

```suggestion
        return self.filter(
            zeros,
            poles,
            gain,
            analog=analog,
            unit=unit,
            **kwargs,
        )
```

File: gwpy/timeseries/timeseries.py [hunk 6]
```diff
@@ -986,7 +990,15 @@ def zpk(self, zeros, poles, gain, analog=True, **kwargs):
 
         >>> data2 = data.zpk([100]*5, [1]*5, 1e-10)
         """
-        return self.filter(zeros, poles, gain, analog=analog, **kwargs)
+
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py:993]: Can you please remove the blank line between the docstring end and the code? There are a few examples in the test modules as well, so please remove those if you can.

```suggestion

```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 7]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+
+        # be within 40% for all values after
+        # numpy allclose formula:
+        # |a-b| <= atol + rtol * |b|
+
+        assert numpy.allclose(
+            hp_asd[eqind0:].value,
+            asd[eqind0:].value,
+            rtol=0.4,
+            atol=0)
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1402]:
```suggestion
            atol=0,
        )
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 8]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+
+        # dont be within 40% for all value before
+        assert not numpy.allclose(
+            hp_asd[:eqind0].value,
+            asd[:eqind0].value,
+            rtol=0.4,
+            atol=0
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1409]:
```suggestion
            atol=0,
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 9]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+
+        # be within 40% for all values before
+        # numpy allclose formula:
+        # |a-b| <= atol + rtol * |b|
+
+        assert not numpy.allclose(
+            lp_asd[eqind0:].value,
+            asd[eqind0:].value,
+            rtol=0.4,
+            atol=0
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1429]:
```suggestion
            atol=0,
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 10]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+        )
+
+        # dont be within 40% for all value after
+        assert numpy.allclose(
+            lp_asd[:eqind0].value,
+            asd[:eqind0].value,
+            rtol=0.4,
+            atol=0
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1437]:
```suggestion
            atol=0,
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 11]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+
+        assert l_inds[-1] <= nf_ind
+        assert r_inds[0] >= nf_ind
+
+        # be within 40% for all values outside nbrhood
+        assert numpy.allclose(
+            notched_asd[l_inds].value,
+            asd[l_inds].value,
+            rtol=0.4,
+            atol=0
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1467]:
```suggestion
            atol=0,
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 12]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+        )
+        assert numpy.allclose(
+            notched_asd[r_inds].value,
+            asd[r_inds].value,
+            rtol=0.4,
+            atol=0
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1473]:
```suggestion
            atol=0,
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 13]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+        )
+
+        # dont be within 40% for all values inside nbrhood
+        assert not numpy.allclose(
+            notched_asd[nf_inds].value,
+            asd[nf_inds].value,
+            rtol=0.4,
+            atol=0
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1481]:
```suggestion
            atol=0,
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 14]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+
+        # be within 40% for all values after
+        # numpy allclose formula:
+        # |a-b| <= atol + rtol * |b|
+
+        assert numpy.allclose(
+            bp_asd[eqind0:eqindn].value,
+            asd[eqind0:eqindn].value,
+            rtol=0.4,
+            atol=0
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1513]:
```suggestion
            atol=0,
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 15]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+        )
+
+        # dont be within 40% for all value before
+        assert not numpy.allclose(
+            bp_asd[:eqind0].value,
+            asd[:eqind0].value,
+            rtol=0.4,
+            atol=0
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1521]:
```suggestion
            atol=0,
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 16]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+        )
+
+        # or after
+        assert not numpy.allclose(
+            bp_asd[eqindn:].value,
+            asd[eqindn:].value,
+            rtol=0.4,
+            atol=0
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1529]:
```suggestion
            atol=0,
```

File: gwpy/timeseries/timeseries.py [hunk 17]
```diff
@@ -986,7 +990,15 @@ def zpk(self, zeros, poles, gain, analog=True, **kwargs):
...
+        return self.filter(
+            zeros,
+            poles,
+            gain,
+            analog=analog,
+            unit=unit,
+            **kwargs
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py:1000]:
```suggestion
            **kwargs,
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 18]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
         utils.assert_quantity_sub_equal(
             gw150914.zpk(*zpk), gw150914.filter(*zpk, analog=True))
 
+    def test_highpass_happy_path(self, gw150914):
+        """Check that passband val are approx equal, stopband are not."""
+
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1387]:
```suggestion
        """Check that passband val are approx equal, stopband are not.
        """
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 19]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+        assert not numpy.allclose(
+            hp_asd[:eqind0].value,
+            asd[:eqind0].value,
+            rtol=0.4,
+            atol=0
+        )
+
+    def test_lowpass_happy_path(self, gw150914):
+        """Check that passband val are approx equal, stopband are not."""
+
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1414]:
```suggestion
        """Check that passband val are approx equal, stopband are not.
        """
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 20]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+        assert numpy.allclose(
+            lp_asd[:eqind0].value,
+            asd[:eqind0].value,
+            rtol=0.4,
+            atol=0
+        )
+
+    def test_notch_happy_path(self, gw150914):
+        """Check passband vals are approx equal, stopband are not."""
+
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1442]:
```suggestion
        """Check passband vals are approx equal, stopband are not.
        """
```

File: gwpy/timeseries/tests/test_timeseries.py [hunk 21]
```diff
@@ -1370,6 +1382,153 @@ def test_zpk(self, gw150914):
...
+        )
+
+        # biggest difference between filtered and unfiltered
+        # should be at closest f to nf=10
+        absd = numpy.abs(notched_asd.value - asd.value)
+        assert numpy.isclose(absd[nf_ind], numpy.max(absd))
+
+    def test_bandpass_happy_path(self, gw150914):
+        """Check that passband val are approx equal, stopband are not."""
+
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py:1491]:
```suggestion
        """Check that passband val are approx equal, stopband are not.
        """
```

**Review by duncanmmacleod — CHANGES_REQUESTED** [review]
Now that I look carefully, some more pedantry.
